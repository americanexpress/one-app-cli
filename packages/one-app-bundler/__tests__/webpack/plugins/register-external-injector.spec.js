/**
 * @jest-environment node
 */

/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const fs = require('node:fs');
const path = require('node:path');
const webpack = require('webpack');
const { merge } = require('webpack-merge');

const { ExternalRegisterPlugin } = require('../../../webpack/plugins/register-external-injector');

const fixturesPath = path.join(__dirname, '__fixtures__');
const buildPath = path.join(fixturesPath, 'build');

const webpackOptions = {
  entry: path.join(fixturesPath, 'index.js'),
  devtool: 'source-map',
  output: {
    path: buildPath,
  },
  plugins: [new ExternalRegisterPlugin('external-a', '1.2.3')],
};

function waitForWebpack(options) {
  // eslint-disable-next-line no-promise-executor-return -- webpack callback
  return new Promise((resolve, reject) => webpack(options, (error, stats) => {
    if (error) { return reject(error); }
    if (stats.hasErrors()) { return reject(stats.toJson().errors); }
    return resolve(stats);
  }));
}

describe('ExternalRegisterPlugin', () => {
  it('should wrap the contents in an IIFE that registers the module in development', async () => {
    expect.assertions(5);

    const outputFileName = 'webpack-test-output-dev.js';
    const options = merge(webpackOptions, {
      mode: 'development',
      output: {
        filename: outputFileName,
      },
    });

    await waitForWebpack(options);

    const fileContents = fs.readFileSync(path.join(buildPath, outputFileName)).toString();

    expect(fileContents.startsWith('/******/ (() => { // webpackBootstrap')).toBe(true);
    expect(fileContents).toContain("const ID = 'external-a';");
    expect(fileContents).toContain('Holocron.registerExternal({ name: "external-a", version: "1.2.3", module: __holocron_external__external_a__1_2_3});');
    expect(fileContents).toContain("console.error('ERROR Registring External \"external-a\"', err)");
    expect(fileContents).toMatchSnapshot();
  });

  it('should wrap the contents in an IIFE that registers the module in production', async () => {
    expect.assertions(1);

    const outputFileName = 'webpack-test-output-prod.js';
    const options = merge(webpackOptions, {
      devtool: false,
      mode: 'production',
      output: {
        filename: outputFileName,
      },
    });

    await waitForWebpack(options);

    const fileContents = fs.readFileSync(path.join(buildPath, outputFileName)).toString();

    expect(fileContents).toContain('Holocron.registerExternal({name:"external-a",version:"1.2.3",module:__holocron_external__external_a__1_2_3})');
  });
});
