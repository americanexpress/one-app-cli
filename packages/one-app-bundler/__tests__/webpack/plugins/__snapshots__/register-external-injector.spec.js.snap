// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalRegisterPlugin should wrap the contents in an IIFE that registers the module in development 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./packages/one-app-bundler/__tests__/webpack/plugins/__fixtures__/external-a.js\\":
/*!***************************************************************************************!*\\\\
  !*** ./packages/one-app-bundler/__tests__/webpack/plugins/__fixtures__/external-a.js ***!
  \\\\***************************************************************************************/
/***/ ((__unused_webpack___webpack_module__, __webpack_exports__, __webpack_require__) => {

__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   ExternalA: () => (/* binding */ ExternalA)
/* harmony export */ });
/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the \\"License\\"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an \\"AS IS\\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const ID = 'external-a';

const ExternalA = () => {
  console.log('hello mundo!');

  return ID;
};


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/define property getters */
/******/ 	(() => {
/******/ 		// define getter functions for harmony exports
/******/ 		__webpack_require__.d = (exports, definition) => {
/******/ 			for(var key in definition) {
/******/ 				if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 					Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 				}
/******/ 			}
/******/ 		};
/******/ 	})();
/******/ 	
/******/ 	/* webpack/runtime/hasOwnProperty shorthand */
/******/ 	(() => {
/******/ 		__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ 	})();
/******/ 	
/******/ 	/* webpack/runtime/make namespace object */
/******/ 	(() => {
/******/ 		// define __esModule on exports
/******/ 		__webpack_require__.r = (exports) => {
/******/ 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 			}
/******/ 			Object.defineProperty(exports, '__esModule', { value: true });
/******/ 		};
/******/ 	})();
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!**********************************************************************************!*\\\\
  !*** ./packages/one-app-bundler/__tests__/webpack/plugins/__fixtures__/index.js ***!
  \\\\**********************************************************************************/
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   ExternalA: () => (/* reexport safe */ _external_a_js__WEBPACK_IMPORTED_MODULE_0__.ExternalA)
/* harmony export */ });
/* harmony import */ var _external_a_js__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./external-a.js */ \\"./packages/one-app-bundler/__tests__/webpack/plugins/__fixtures__/external-a.js\\");
/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the \\"License\\"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an \\"AS IS\\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */



})();

;try {
    Holocron.registerExternal({ name: \\"external-a\\", version: \\"1.2.3\\", module: __holocron_external__external_a__1_2_3});
} catch (err) {
    console.error('ERROR Registring External \\"external-a\\"', err)
}/******/ })()
;
//# sourceMappingURL=webpack-test-output-dev.js.map"
`;
